pkg/registrar/bastion.go

```diff
@@ -0,0 +1,138 @@
+package registrar
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/giantswarm/microerror"
+	"github.com/go-logr/logr"
+	clouddns "google.golang.org/api/dns/v1"
+	capg "sigs.k8s.io/cluster-api-provider-gcp/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+//counterfeiter:generate . BastionsClient
+type BastionsClient interface {
+	GetBastionIPList(ctx context.Context, cluster *capg.GCPCluster) ([]string, error)
+	AddFinalizerToBastions(ctx context.Context, cluster *capg.GCPCluster) error
+	RemoveFinalizerFromBastions(ctx context.Context, cluster *capg.GCPCluster) error
+}
+
+type Bastion struct {
+	baseDomain     string
+	bastionsClient BastionsClient
+	dnsService     *clouddns.Service
+}
+
+func NewBastion(baseDomain string, bastionsClient BastionsClient, dnsService *clouddns.Service) *Bastion {
+	return &Bastion{
+		baseDomain:     baseDomain,
+		bastionsClient: bastionsClient,
+		dnsService:     dnsService,
+	}
+}
+
+func (r *Bastion) Register(ctx context.Context, cluster *capg.GCPCluster) error {
+	logger := r.getLogger(ctx)
+	err := r.bastionsClient.AddFinalizerToBastions(ctx, cluster)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
```

Comment:
I don't think we receive events for GCPMachines, so isn't this dangerous? It would mean that if a GCPMachine is deleted it'll be stuck until something triggers an event on the GCPCluster and the finalizer is removed.